### ISSUE TITLE
-st-extends specificity and validations

### ISSUE THREAD

**Issue by tomrav** [wix/stylable#355]
### Existing situation
When using an extended class we can run into three different use-cases.

1. Using a class in the same scope
2. Accessing a class name through a pseudo-element (no explicit import)
3. Importing a class and customizing it in a different stylesheet

Example:
```css
/* button.st.css */
.label {} /* case 1 */
```

```css
/* app.st.css */
:import {
    -st-from: "./button.st.css";
    -st-default: Button;
    -st-named: label;
}

.myButton {
    -st-extends: Button;
}
.myButton::label {} /* case 2 */

.root .label {} /* case 3 */
```

Output CSS:
```css
.button__label {} /* case 1 */
.app__myButton.button__root .button__label {} /* case 2 */
.app__root .button__label {} /* case 3 */
```

In the 2nd case, the `.button__root` class was added to signify the inheritance. This leads to two main impacts:
1. Increased specificity, which causes the 2nd use case to beat the 3rd
2. Added validation, this selector will not be applied to just any element, but rather only to those that have their root set by the `Button` component

### Proposed change
We can simplify selector complexity, reducing size and increasing readability in such cases by **not** adding the additional root class upon `extend`.

This change causes the 2nd and 3rd use-cases above to have equal specificity, causing their order in the document to determine their strength. In order to still offer validations when extending we can add additional selectors in `dev` time to provide visual indication for using a class on an incorrect type.

Example:
```css
:import {
    -st-from: "./button.st.css";
    -st-default: Button;
}

.myButton {
    -st-extends: Button;
}

.myOtherButton {
    -st-extends: div; /* no way to extend native nodes at the moment */ 
}
```

Dev-time CSS can offer runtime validation rendered to the DOM:
```css
.app__myButton:not(.button__root)::before {
    content: "INVALID CSS CLASS ASSIGNMENT of 'app::myButton'" !important; 
}
.app__myButton {}
 
.app__myOtherButton:not(div)::before {
    content: "INVALID CSS CLASS ASSIGNMENT of 'app::myOtherButton'" !important; 
} 
.app__myOtherButton {}
```

**Comment by barak007** [comment]
This is a breaking change. consider this case

a.st.css
```css  
.root {color: #000}
.root:hover {color: #001}
```
site.st.css
```css
:import {
    -st-from: "./a.st.css";
    -st-default: A;
}
.root {
    -st-extends: A;
    color: #002;
}
```

output
```css
.a__root {color: #000}
.a__root:hover {color: #001}
.site__root.a__root {color: #002}
```

As stylable now. The site root will be stronger then the hover state of the base.

There are two things that I can say about it and I don't know if they are good or bad.
1. It's bad because it is only working for one specificity level.
2. Without the scoping the override strength is purely determent by the order of the style sheets.

**Comment by idoros** [comment]
I believe that this change is good, because it simplifies the output, bringing it closer to the source. And because the addition of class name only slightly protects the extending context, making it a confusing feature.

I think we should simply remove it. We guarantee the order, and its up to the developer to at list match the specificity of the extended context.

In the future when CSS custom selectors will give us control of specificity , we can offer a way to absolutely override extended context.

**Comment by tomrav** [comment]
Resolved in #364 and merged to branch `scoped-stylable` for upcoming major release.
